readthedocs/api/v3/views.py

```diff
@@ -384,30 +384,117 @@ def create(self, request, **kwargs):  # pylint: disable=arguments-differ
+        project = Notification.objects.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(Project),
+            attached_to_id__in=Subquery(
+                AdminPermission.projects(
+                    self.request.user, admin=True, member=False
+                ).values("id")
+            ),
+        )
+        organization = Notification.objects.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(Organization),
+            attached_to_id__in=Subquery(
+                AdminPermission.organizations(
+                    self.request.user, admin=True, member=False
+                ).values("id")
+            ),
+        )
+
+        # Return all the notifications related to this user attached to:
+        # User, Project and Organization models where the user is admin.
+        return user | project | organization
+
+
+# TODO: verify permissions over all ``**/notifications/`` endpoints.
+class NotificationsProjectViewSet(
     APIv3Settings,
     NestedViewSetMixin,
     ProjectQuerySetMixin,
     FlexFieldsMixin,
-    ReadOnlyModelViewSet,
+    ListModelMixin,
+    RetrieveModelMixin,
+    UpdateMixin,
+    UpdateModelMixin,
+    GenericViewSet,
 ):
     model = Notification
     lookup_field = "pk"
     lookup_url_kwarg = "notification_pk"
     serializer_class = NotificationSerializer
     queryset = Notification.objects.all()
-    # filterset_class = BuildFilter
+    filterset_class = NotificationFilter
 
-    # http://chibisov.github.io/drf-extensions/docs/#usage-with-generic-relations
     def get_queryset(self):
-        queryset = self.queryset.filter(
-            attached_to_content_type=ContentType.objects.get_for_model(Build)
-        )
+        project = self._get_parent_project()
+        return project.notifications.all()
 
-        # TODO: make sure if this particular filter should be applied here or somewhere else.
-        return queryset.filter(
-            attached_to_id=self.kwargs.get("parent_lookup_build__id")
-        )
+
+class NotificationsBuildViewSet(
+    APIv3Settings,
+    NestedViewSetMixin,
+    ProjectQuerySetMixin,
+    FlexFieldsMixin,
+    ListModelMixin,
+    RetrieveModelMixin,
+    UpdateMixin,
+    UpdateModelMixin,
+    GenericViewSet,
+):
+    model = Notification
+    lookup_field = "pk"
+    lookup_url_kwarg = "notification_pk"
+    serializer_class = NotificationSerializer
+    queryset = Notification.objects.all()
+    filterset_class = NotificationFilter
+
+    def get_queryset(self):
+        build = self._get_parent_build()
+        return build.notifications.all()
+
+
+# TODO: should we create different endpoints?
+# - List all notifications related to the user (User, Project, Organization): /api/v3/notifications/
+# - List all notifications related to a particular Build: /api/v3/projects/docs/builds/12345/notifications/
+# - Update notification: /api/v3/notifications/123/
```

Comment:
What are you thinking here for marking notifications read, we will always PATCH against `/api/v3/notifications/123` instead of `/api/v3/projects/456/notifications/123`?

I think this makes sense, but also there probably isn't much difference either. The front end will know which URL to use.

Reply:
He, this was a "brain dump" while I was making decisions about the correct pattern. It makes no sense anymore and these comments need to be deleted.

We will follow what the description of this PR says. That is hitting `PATCH /api/v3/projects/456/notifications/123`, since it follows the same pattern we have for the rest of the resources.

That URL is returned under `_links` anyways, so the front-end will use that field.